enhancements/installer/feature-sets.md

```diff
@@ -0,0 +1,224 @@
+			allErrs = append(allErrs, field.Forbidden(field.NewPath("foo", "bar"), errMsg))
+		}
+	}
+```
+
+Package `types` defines the install config and represents the API offered by the Installer. Most notably this
+API is consumed/used by hive. For this reason, it is important that feature sets be part of the install config.
+
+#### CLI Flag (future work)
+
+On the other hand, the install config is an asset and gated features could potentially be exercised
+before the install config is validated (or perhaps the feature is completely independent from the install config). For this reason, it would
+be a good idea to provide a flag:
+
+```shell
+$ openshift-install create cluster --feature-set TechPreviewNoUpgrade
+```
+
+This flag, if provided, would also populate the value in the install config; and conflicts between the flag and install config
+would throw an error.
+
+While there are clear use cases for a flag, there are no immediate use cases and the implementation of the Installer (with
+its separation between command-line flags and assets) does not lead to a clear-cut design. Implementing this functionality
+is certainly possible, but I propose we defer implementation for this until it is determined to be necessary.
+
+#### openshift-install explain
+
+`explain` provides a command-line reference for the install config:
+
+```shell
+$ openshift-install explain installconfig | tail -n 6
+    pullSecret <string> -required-
+      PullSecret is the secret to use when pulling images.
+
+    sshKey <string>
+      SSHKey is the public Secure Shell (SSH) key to provide access to instances.
+```
+
+`explain` should handle feature-gated fields appropriately, either by skipping them (not printing them) or clearly marking
+them as gated features. The initial implementation will simply include
+comments denoting fields that are feature gated.
+
+### Risks and Mitigations
+
+I am not aware of any significant risks to introducing this feature. 
+
+### Drawbacks
+
+I am not aware of any significant drawbacks to introducing this feature.
+
+## Design Details
+
+### Open Questions [optional]
+
+N/A
+
+### Test Plan
+
+Unit testing the new tech preview functionality and observing existing e2e tests should be sufficient.
+
+### Graduation Criteria
+
+n/a
+
+#### Dev Preview -> Tech Preview
+
+#### Tech Preview -> GA
+
+#### Removing a deprecated feature
+
+
+### Upgrade / Downgrade Strategy
+
+Upgrades will be blocked based on the feature set. There is
+discussion about whether it makes sense to block a cluster
+upgrade based on Installer-only features, i.e. if the changes
+are only in the Installer and don't affect the cluster, why
+block upgrades?
+
+The upgrade status of the cluster is based on the feature set,
+so the current implementation would have the ability to accomodate
+an Installer-only tech preview if we were able to add a new feature
+set for that purpose. For discussion: if we added an `InstallerTechPreview` or `TechPreviewInstallOnly` feature set,
```

Comment:
I'm not in favor of adding this option.  If a feature is complete enough that we're ready to promise upgrades, then the feature is complete enough to be GA because it means that nothing can change about the user's interaction with the feature and it must produce reliable output.  Even from the installer since it produces lasting impacts on production clusters.

Reply:
I agree we should not allow a middle ground. This is a forcing function which pushes us not to leave something in an experimental state for longer than necessary before committing ourselves to fully support it.